### Introduction

This project examines the tweet archive for WeRateDogs to draw insights from the dataset regarding the ratings of dogs. The Gather, Assess and Clean process is followed with the aim of answering the following questions:

- What type of dogs have the highest ratings?
- What dog stage has the highest ratings?
- What type of dogs are the most popular in terms of retweet count and favorite count?

## Gather

In [1]:
import json
import numpy as np
import os
import pandas as pd
import requests
import tweepy
from timeit import default_timer as timer
from tweepy import OAuthHandler

In [2]:
# Load twitter archive file
twitter_archive_enhanced = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# Download image-predictions file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open('image_predictions.tsv', mode='wb') as file:
    file.write(response.content)

In [4]:
# Import image predictions file
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

In [5]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [6]:
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = twitter_archive_enhanced.tweet_id.values
len(tweet_ids)

2356

In [7]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
if not os.path.exists('tweet_json.txt'):
    count = 0
    fails_dict = {}
    start = timer()
    # Save each tweet's returned JSON as a new line in a .txt file
    with open('tweet_json.txt', 'w') as outfile:
        # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    end = timer()
    print(end - start)
    print(fails_dict)


In [8]:
# DELETE
# Temp - save fails_dict to file for working with
'''
import csv
fails_list = []
for key in fails_dict.keys():
    fails_list.append(key)
fails_list
with open('fails_dict.csv', 'w') as outfile:
    wr = csv.writer(outfile, dialect='excel')
    wr.writerow(fails_list)
'''

"\nimport csv\nfails_list = []\nfor key in fails_dict.keys():\n    fails_list.append(key)\nfails_list\nwith open('fails_dict.csv', 'w') as outfile:\n    wr = csv.writer(outfile, dialect='excel')\n    wr.writerow(fails_list)\n"

In [9]:
# DELETE
# Working - uncomment out if necessary
'''
failures = []
with open('fails_dict.csv', 'r') as f:
    reader = csv.reader(f)
    for line in reader:
        failures.append(line)
'''

"\nfailures = []\nwith open('fails_dict.csv', 'r') as f:\n    reader = csv.reader(f)\n    for line in reader:\n        failures.append(line)\n"

In [10]:
# Load extended tweet data
tweets = []
with open('tweet_json.txt') as json_file:
    for line in json_file:
        tweets.append(json.loads(line))

# Place extended tweet data into a dataframe
extended_data = pd.DataFrame(tweets)

## Assess

In [11]:
# View column types and missing data in twitter_archive_enhanced
twitter_archive_enhanced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
# DETERMINE IF TO KEEP
# Visually assess data in archive
twitter_archive_enhanced.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
58    880935762899988482                    NaN                  NaN   
1393  700029284593901568                    NaN                  NaN   
1915  674318007229923329                    NaN                  NaN   
2052  671486386088865792                    NaN                  NaN   
115   870374049280663552                    NaN                  NaN   

                      timestamp  \
58    2017-06-30 23:47:07 +0000   
1393  2016-02-17 18:49:22 +0000   
1915  2015-12-08 20:01:55 +0000   
2052  2015-12-01 00:30:04 +0000   
115   2017-06-01 20:18:38 +0000   

                                                 source  \
58    <a href="http://twitter.com/download/iphone" r...   
1393  <a href="http://twitter.com/download/iphone" r...   
1915  <a href="http://twitter.com/download/iphone" r...   
2052  <a href="http://twitter.com/download/iphone" r...   
115   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
58    This is Louis. He's crossing. It's a big deal....                  NaN   
1393  This is Coops. His ship is taking on water. So...                  NaN   
1915  This is Lennon. He's in quite the predicament....                  NaN   
2052  This is Malcolm. He just saw a spider. 10/10 h...                  NaN   
115   This is Zoey. She really likes the planet. Wou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
58                         NaN                        NaN   
1393                       NaN                        NaN   
1915                       NaN                        NaN   
2052                       NaN                        NaN   
115                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
58    https://twitter.com/dog_rates/status/880935762...                13   
1393  https://twitter.com/dog_rates/status/700029284...                10   
1915  https://twitter.com/dog_rates/status/674318007...                 8   
2052  https://twitter.com/dog_rates/status/671486386...                10   
115   https://twitter.com/dog_rates/status/870374049...                13   

      rating_denominator     name doggo floofer  pupper puppo  
58                    10    Louis  None    None    None  None  
1393                  10    Coops  None    None    None  None  
1915                  10   Lennon  None    None  pupper  None  
2052                  10  Malcolm  None    None    None  None  
115                   10     Zoey  None    None    None  None

In [13]:
# Check for duplicated tweets
len(twitter_archive_enhanced[twitter_archive_enhanced.tweet_id.duplicated()])

0

In [14]:
# Check for inclusion of retweets
sum(~twitter_archive_enhanced.retweeted_status_id.isnull())

181

There are 181 tweets that are retweets and need to be removed.

In [15]:
# Check values in rating_denominator - should be 10
twitter_archive_enhanced.rating_denominator.value_counts().sort_index()

0         1
2         1
7         1
10     2333
11        3
15        1
16        1
20        2
40        1
50        3
70        1
80        2
90        1
110       1
120       1
130       1
150       1
170       1
Name: rating_denominator, dtype: int64

In [16]:
# Check for incorrect rating_denominator values
incorrect_denominators = twitter_archive_enhanced[twitter_archive_enhanced['rating_denominator'] != 10]
list(incorrect_denominators['tweet_id'])

[835246439529840640,
 832088576586297345,
 820690176645140481,
 810984652412424192,
 775096608509886464,
 758467244762497024,
 740373189193256964,
 731156023742988288,
 722974582966214656,
 716439118184652801,
 713900603437621249,
 710658690886586372,
 709198395643068416,
 704054845121142784,
 697463031882764288,
 686035780142297088,
 684225744407494656,
 684222868335505415,
 682962037429899265,
 682808988178739200,
 677716515794329600,
 675853064436391936,
 666287406224695296]

Some incorrect denominators were found:

- 835246439529840640 should be 13 numerator and 10 denominator
- 832088576586297345 does not contain a rating and should be dropped
- 810984652412424192 does not contain a rating and should be dropped
- 775096608509886464 is a retweet and will be deleted
- 740373189193256964 should be 14 numerator 10 denominator
- 722974582966214656 should be 13 numerator and 10 denominator
- 716439118184652801 should be 11 numerator and 10 denominator
- 686035780142297088 does not contain a rating and should be dropped
- 682962037429899265 should be 10 numerator and 10 denominator
- 682808988178739200 does not contain a rating and should be dropped
- 666287406224695296 should be 9 numerator and 10 denominator

Others from the list are accurate although they do not have 10 as a denominator.

In [17]:
# Examine data types and mising data in image_predictions
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [18]:
# DETERMINE IF TO KEEP
image_predictions.sample(10)

tweet_id                                          jpg_url  \
317   671744970634719232  https://pbs.twimg.com/media/CVKEfMKWoAAR-Ud.jpg   
1011  709449600415961088  https://pbs.twimg.com/media/Cdh4pgAW0AEKJ_a.jpg   
340   672248013293752320  https://pbs.twimg.com/media/CVROAIfWsAECA5t.jpg   
1035  711743778164514816  https://pbs.twimg.com/media/CeCfMPDW0AAAEUj.jpg   
478   675334060156301312  https://pbs.twimg.com/media/CV9EvZNUwAAgLCK.jpg   
1608  800751577355128832  https://pbs.twimg.com/media/CxzXOyBW8AEu_Oi.jpg   
608   680070545539371008  https://pbs.twimg.com/media/CW-dU34WQAANBGy.jpg   
1098  720389942216527872  https://pbs.twimg.com/media/Cf9W1J-UMAErahM.jpg   
772   689557536375177216  https://pbs.twimg.com/media/CZHM60BWIAA4AY4.jpg   
714   685532292383666176  https://pbs.twimg.com/media/CYN_-6iW8AQhPu2.jpg   

      img_num                p1   p1_conf  p1_dog                p2   p2_conf  \
317         1          ice_bear  0.251193   False               ram  0.213839   
1011        2       Maltese_dog  0.780187    True    Dandie_Dinmont  0.074429   
340         1     Irish_terrier  0.413173    True          Airedale  0.335616   
1035        1  Lakeland_terrier  0.459515    True  miniature_poodle  0.219661   
478         2          Pembroke  0.773135    True          Cardigan  0.116810   
1608        2    cocker_spaniel  0.771984    True  miniature_poodle  0.076653   
608         1         earthstar  0.127701   False          Shih-Tzu  0.121811   
1098        1          Pembroke  0.873977    True          Cardigan  0.043339   
772         1        Eskimo_dog  0.169482    True    Siberian_husky  0.161655   
714         1        white_wolf  0.318524   False             dingo  0.215436   

      p2_dog               p3   p3_conf  p3_dog  
317    False       Arctic_fox  0.081551   False  
1011    True  Norfolk_terrier  0.033776    True  
340     True       toy_poodle  0.027952    True  
1035    True  standard_poodle  0.130189    True  
478     True             chow  0.039036    True  
1608    True       toy_poodle  0.039618    True  
608     True           bubble  0.117820   False  
1098    True       Eskimo_dog  0.019197    True  
772     True            dingo  0.154414   False  
714    False           collie  0.095805    True

In [19]:
# Check column types and missing values in extended_data
extended_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2339 entries, 0 to 2338
Data columns (total 32 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2339 non-null object
display_text_range               2339 non-null object
entities                         2339 non-null object
extended_entities                2065 non-null object
favorite_count                   2339 non-null int64
favorited                        2339 non-null bool
full_text                        2339 non-null object
geo                              0 non-null object
id                               2339 non-null int64
id_str                           2339 non-null object
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null object
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null obj

In [20]:
# DETERMINE IF TO KEEP
extended_data.head(5)

contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   
2         None        None  Mon Jul 31 00:18:03 +0000 2017           [0, 121]   
3         None        None  Sun Jul 30 15:58:51 +0000 2017            [0, 79]   
4         None        None  Sat Jul 29 16:00:24 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           37783   
1  {'media': [{'id': 892177413194625024, 'id_str'...           32453   
2  {'media': [{'id': 891815175371796480, 'id_str'...           24437   
3  {'media': [{'id': 891689552724799489, 'id_str'...           41113   
4  {'media': [{'id': 891327551943041024, 'id_str'...           39326   

   favorited                                          full_text   geo  ...  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None  ...   
1      False  This is Tilly. She's just checking pup on you....  None  ...   
2      False  This is Archie. He is a rare Norwegian Pouncin...  None  ...   
3      False  This is Darla. She commenced a snooze mid meal...  None  ...   
4      False  This is Franklin. He would like you to stop ca...  None  ...   

   quoted_status quoted_status_id quoted_status_id_str  \
0            NaN              NaN                  NaN   
1            NaN              NaN                  NaN   
2            NaN              NaN                  NaN   
3            NaN              NaN                  NaN   
4            NaN              NaN                  NaN   

   quoted_status_permalink retweet_count  retweeted retweeted_status  \
0                      NaN          8233      False              NaN   
1                      NaN          6084      False              NaN   
2                      NaN          4027      False              NaN   
3                      NaN          8384      False              NaN   
4                      NaN          9089      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [21]:
# Find number of retweets in extended_data
len(extended_data[~extended_data['retweeted_status'].isnull()])

167

There are 167 tweets that are retweets and will need to be removed from the data.

In [22]:
# Check for duplication between twitter_archive_enhanced and image_predictions
all_columns = pd.Series(list(twitter_archive_enhanced) + list(image_predictions))
all_columns[all_columns.duplicated()]

17    tweet_id
dtype: object

Only the tweet_id column was duplicated, which we can use as the link between to two tables.

#### Quality
##### `twitter_archive_enhanced` table
- data missing such as retweet_count and favorite_count
- tweet_id is an int (should be str)
- presence of retweets (retweeted_status_id is non-null)
- rating_numerator and rating_denominator columns are ints (should be floats)
- several incorrect ratings to be corrected

##### `image_predictions` table
- tweet_id is an int (should be str)
- dog names in p1, p2, p3 columns inconsistent - upper and lowercase names, inclusion of underscores etc.
- predictions include non-dog objects which should be removed prior to analysis

##### `extended_data` table
- presence of retweets (retweeted_status is non-null)

#### Tidiness
- `twitter_archive_enhanced` contains multiple columns for the dog stage which should be one column with the dog stage as the variable
- `image_predictions` has multiple predictions per row

## Clean

In [23]:
# Make copies of each dataframe for cleaning
twitter_archive_enhanced_clean = twitter_archive_enhanced.copy()
image_predictions_clean = image_predictions.copy()
extended_data_clean = extended_data.copy()

### Missing Data

#### `extended_info`: favorite_count and retweet_count need to be joined to `twitter_archive_enhanced`.

##### Define
- Drop the unwanted columns from `extended_data_clean` so just id, favorite_count and retweet_count remain
- Rename id to tweet_id in `extended_data_clean`
- Merge `extended_data_clean` with `twitter_archive_enhanced_clean` on the tweet_id column, keeping only the rows (tweet_id) that are present in both dataframes

##### Code

In [24]:
# Remove unwanted columns
columns = ['id', 'favorite_count', 'retweet_count']
extended_data_clean = extended_data_clean[columns]

# Rename id column
extended_data_clean = extended_data_clean.rename(columns={'id':'tweet_id'})
extended_data_clean.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           37783           8233
1  892177421306343426           32453           6084
2  891815181378084864           24437           4027
3  891689557279858688           41113           8384
4  891327558926688256           39326           9089

In [25]:
# Merge twitter_archive_enhanced_clean and extended_data_clean
twitter_archive_enhanced_clean = pd.merge(twitter_archive_enhanced_clean, extended_data_clean, on='tweet_id', how='inner')

##### Test

In [26]:
# Check that retweet_count and favorite_count columns have been added and there are 2339 rows
twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2339 entries, 0 to 2338
Data columns (total 19 columns):
tweet_id                      2339 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2339 non-null object
source                        2339 non-null object
text                          2339 non-null object
retweeted_status_id           167 non-null float64
retweeted_status_user_id      167 non-null float64
retweeted_status_timestamp    167 non-null object
expanded_urls                 2280 non-null object
rating_numerator              2339 non-null int64
rating_denominator            2339 non-null int64
name                          2339 non-null object
doggo                         2339 non-null object
floofer                       2339 non-null object
pupper                        2339 non-null object
puppo                         2339 non-null object
favorite_count                23

### Tidiness

#### `twitter_archive_enhanced` contains multiple columns for the dog stage.

##### Define
- Create a column in `twitter_archive_enhanced` called dog_stage that holds the concatenation of each value in the doggo, floofer, pupper and puppo columns.
- Extract the correct dog stage from the dog_stage column and update
- For entries with more than one correct dog, record these as 'mixed'

##### Code

In [27]:
# Add dog_stage column and populate with concatenated values for dog stages
twitter_archive_enhanced_clean['dog_stage'] = twitter_archive_enhanced_clean['doggo'] + \
    twitter_archive_enhanced_clean['floofer'] + twitter_archive_enhanced_clean['pupper'] + \
    twitter_archive_enhanced_clean['puppo']

In [28]:
# Check values present in dog_stage column
twitter_archive_enhanced_clean['dog_stage'].value_counts()

NoneNoneNoneNone        1961
NoneNonepupperNone       244
doggoNoneNoneNone         82
NoneNoneNonepuppo         29
doggoNonepupperNone       12
NoneflooferNoneNone        9
doggoNoneNonepuppo         1
doggoflooferNoneNone       1
Name: dog_stage, dtype: int64

In [29]:
# Function to extract dog_stage
def get_dog_stage(current):
    '''Extract dog_stage from passed string'''
    if current == 'NoneNoneNoneNone':
        return 'None'
    elif current == 'doggoNoneNoneNone':
        return 'doggo'
    elif current == 'NoneflooferNoneNone':
        return 'floofer'
    elif current == 'NoneNonepupperNone':
        return 'pupper'
    elif current == 'NoneNoneNonepuppo':
        return 'puppo'
    else:
        return 'mixed'

In [30]:
# Update dog_stage column with call to get_dog_stage()
twitter_archive_enhanced_clean['dog_stage'] = twitter_archive_enhanced_clean['dog_stage'].apply(get_dog_stage)
# Drop individual dog style columns
twitter_archive_enhanced_clean = twitter_archive_enhanced_clean.drop(['doggo', 'puppo', 'pupper', 'floofer'], axis=1)

##### Test

In [31]:
# Confirm that column has been updated correctly
twitter_archive_enhanced_clean['dog_stage'].value_counts()

None       1961
pupper      244
doggo        82
puppo        29
mixed        14
floofer       9
Name: dog_stage, dtype: int64

In [33]:
# Confirm that old dog stage columns have been dropped
twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2339 entries, 0 to 2338
Data columns (total 16 columns):
tweet_id                      2339 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2339 non-null object
source                        2339 non-null object
text                          2339 non-null object
retweeted_status_id           167 non-null float64
retweeted_status_user_id      167 non-null float64
retweeted_status_timestamp    167 non-null object
expanded_urls                 2280 non-null object
rating_numerator              2339 non-null int64
rating_denominator            2339 non-null int64
name                          2339 non-null object
favorite_count                2339 non-null int64
retweet_count                 2339 non-null int64
dog_stage                     2339 non-null object
dtypes: float64(4), int64(5), object(7)
memory usage: 310.6+ KB


#### `image_predictions` has multiple predictions per row.

##### Define
- Create separate dataframes from `image_predictions_clean` for p1, p2 and p3 data and keep only the columns for that probability (`p1_df`, `p2_df`, `p3_df`)
- For each new dataframe, create a column that holds the p_order value (e.g. 1 for p1)
- Rename p1 to p_type
- Repeat for p2 and p3
- Join `p2` to `p1`
- Join `p3` to `p1`
- Store `p1` as `image_predictions_clean`

##### Code

##### Test

### Quality

#### `twitter_archive_enhanced`: tweet_id is an int

##### Define
- Convert tweet_id column to str using .astype()

##### Code

##### Test

#### `image_predictions`: tweet_id is an int

##### Define
- Convert tweet_id column to str using .astype()

##### Code

##### Test

#### `twitter_archive_enhanced`: presence of retweets

##### Define
- Remove rows where retweet_status_id != null

##### Code

##### Test

#### `extended_data`: presence of retweets

##### Define
- Will be removed from the merged twitter_archive_enhanced table when it is cleaned above.

##### Code

##### Test

#### rating_numerator and rating_denominator columns are ints

##### Define
- Convert rating_numerator and rating_denominator columns to float using .astype()

##### Code

##### Test

#### Incorrect ratings (numerators and denominators)

##### Define
- Write a regex to find the numerator and denominator from the text column
- Save as floats to numerator and denominator columns
- Test against some of the known incorrect values to see if the algorithm improves on original algorithm

##### Code

##### Test

#### Inconsistent format for dog names in p columns of `image_predictions`.

##### Define
- Replace '_' with ' ' in p_type column of `image_predictions_clean`
- Capitalise each word in p_type column of `image_predictions_clean`

##### Code

##### Test

#### Predictions include non-dog objects in `image_predictions`.

##### Define
- Drop rows where p_dog == False in `image_predictions_clean`

##### Code

##### Test

## Storage

(store each table as a csv file with main file called tiwtter_archive_master.csv)

## Analysis

(make at least 3 insights)